### ISSUE TITLE
Make throwFacebookResponseStatusExceptionIfNecessary in DefaultFacebookClient public???

### ISSUE THREAD

**Issue by slumx** [restfb/restfb#80]
When doing batch requests it would be nice to be able to call throwFacebookResponseStatusExceptionIfNecessary, so you don't have to duplicate all the code which is already written.

The downside of this is that you can't use the abstraction of the FacebookClient interface anymore. Maybe the exception handling could be moved to a utility class which could be public accessible?

If know it's a lot to ask. I would be happy to help if we could could agree on the right way to implement this feature...

Here is a quite dirty hack, but it should work:

```
class RestFbExceptionHelper extends DefaultFacebookClient {
	
	protected void throwFacebookResponseStatusExceptionIfNecessary(String json, Integer httpStatusCode) {
		super.throwFacebookResponseStatusExceptionIfNecessary(json, httpStatusCode);
	}
}
```

```
public class RestFbExceptionUtils {
	
	private static final RestFbExceptionHelper restFbExceptionHelper = new RestFbExceptionHelper();
	
	public static void throwFacebookResponseStatusExceptionIfNecessary(String json, Integer httpStatusCode) {
		restFbExceptionHelper.throwFacebookResponseStatusExceptionIfNecessary(json, httpStatusCode);
	}
}
```

When processing larger results the error checking can also take quite some time up to 150ms (depends on hardware ;-)) this could maybe be optimized by adding a little string test in throwFacebookResponseStatusExceptionIfNecessary after testing for "{" something like:

```
int subStrEnd = Math.min(50, json.length());
if (json.substring(0, subStrEnd).indexOf("\"error\"") == -1) {
	return; //no need to parse json
}
```

This gives a constant performance below 1ms no matter how large the json result is.

**Comment by nbartels** [comment]
It's possible to implement your own exception handling class. This should solve this issue.

**Comment by slumx** [comment]
I would still need to do all the parsing and checking done inthrowFacebookResponseStatusExceptionIfNecessary myself??

**Comment by nbartels** [comment]
yes and no, you may extend the default class and use the super method. Some code is encapsulated in methods now. Perhaps you have a look at the new code.

Can you please tell me what's your use case and why you need to change the parsing?

**Comment by slumx** [comment]
I would like the parsing to be the same. But when I'm making a batchRequest i need to test each individual response for errors and I can't do this now because throwFacebookResponseStatusExceptionIfNecessary is protected. Maybe a parse method as in fetchXxx without the fetch fase would be even better because we could then avoid the manual handling of the json altogether. 

How about the performance optimization I suggested in throwFacebookResponseStatusExceptionIfNecessary 😊

**Comment by nbartels** [comment]
Okay, I see. I have to think about the problem and how this can be handled in a generic way :smile:

Your performance optimization is already in your code for a long time. I moved it to a special method, so the duplicate code ist no more necessary. This can be found in the current master branch. 
https://github.com/restfb/restfb/blob/master/source/library/com/restfb/BaseFacebookClient.java#L213

I re-open this issue and leave it open until I have a solutions that works in your case, too.

**Comment by slumx** [comment]
Thanks. I must have browsed through BaseFacebookClient to quickly,  the optimization is of cause right in there 😊

**Comment by nbartels** [comment]
@slumx, I think the latest change regarding this issue should solve your problem. The complete error conversion is moved to a new class with interface. So you may extend this class and use the provided methods. 

Perhaps you can have a look at the commit above. These changes are included in RestFB 2.0.0, so you have to check the dev-2.0 branch to see the latest version.

If you have any question don't hesitate to ask me. As soon as this version is published we publish a section on our website that belongs to this change.

**Comment by slumx** [comment]
Yes that looks much better, thanks...
